3.analyze_data/utils/isolation_forest_data_feature_importance.py

```diff
@@ -0,0 +1,148 @@
+        Aggregates and then returns feature importances per tree.
+
+        Parameters
+        _tree_obj: Object for traversing tree.
+        _leaf_id: Leaf node for the sample in the tree.
+        _sample_idx: Sample location in the dataframe.
+        """
+
+        node_id = 0  # Start at the root node
+        depth = 0
+        feature_importances = defaultdict(list)
+        morphology_features = self._morphology_data.iloc[_sample_idx].copy()
+
+        while node_id != _leaf_id:
+            feature_idx = _tree_obj.feature[node_id]
+
+            if feature_idx >= 0:  # Ignore leaf nodes (-2)
+                """
+                Samples are often easier to seperate with less data.
+                Therefore, earlier features have greater importances.
+                """
+                feature_importances[self._morphology_data.columns[feature_idx]].append(
+                    1 / (1 + depth)
+                )
+
+            if morphology_features.iloc[feature_idx] <= _tree_obj.threshold[node_id]:
+                node_id = _tree_obj.children_left[node_id]
+            else:
+                node_id = _tree_obj.children_right[node_id]
+
+            depth += 1
+
+        """
+        Gives a balanced view of expected feature importance in the tree.
+        As opposed to the total contribution, which increases with the (same) features.
+        Since features are choosen randomly, dividing feature importances
+        by the depth weights each feature by how soon the data could be isolated.
+        """
+
+        return {
+            _sample_idx: {
+                feature: (sum(importances) / len(importances)) / (depth + 1)
+                for feature, importances in feature_importances.items()
+            }
+        }
+
+    def compute_isoforest_feature_importances(self) -> pd.DataFrame:
+        # Computes and returns the importance for all trees and samples
+
+        isotree_importances = Parallel(n_jobs=-1)(
+            delayed(self.save_tree_feature_importances)(
+                _tree_obj=estimator.tree_, _leaf_id=leaf_id, _sample_idx=sample_idx
+            )
+            for estimator in self._estimators
+            for sample_idx, leaf_id in enumerate(
+                estimator.tree_.apply(self._morphology_data.values.astype(np.float32))
+            )
+        )
+
+        isoforest_importances = {}
+        unique_feats = set()
+
+        sample_isoforest_importances = {
+            sample: defaultdict(list)
+            for sample in range(self._morphology_data.shape[0])
+        }
+
+        # Converting data from list of tree dictionaries to dictionary of samples
+        for isotree in isotree_importances:
+            for sample, feature_importances in isotree.items():
+                for feature, importance in feature_importances.items():
+                    sample_isoforest_importances[sample][feature].append(importance)
+
+        # Averaging feature importances per sample and feature across trees
```

Comment:
I understand why it useful to calculate average feature importances across trees, but I don't understand why it's useful to calculate it per sample.

Reply:
looking below, it actually doesn't look like you average feature importances, but only calculate per sample?